core/src/worker/activities.rs

```diff
@@ -514,6 +572,54 @@ pub(crate) struct TrackedPermittedTqResp {
     pub resp: PollActivityTaskQueueResponse,
 }
 
+struct GracefulShutdown {
+    once: TakeCell<GSState>,
+}
+struct GSState {
+    start_tasks_stream_complete: CancellationToken,
+    cancels_tx: UnboundedSender<PendingActivityCancel>,
+    grace_period: Duration,
+    outstanding: OutstandingActMap,
+}
+impl GracefulShutdown {
+    fn new(
+        grace_period: Duration,
+        start_tasks_stream_complete: CancellationToken,
+        cancels_tx: UnboundedSender<PendingActivityCancel>,
+        outstanding: OutstandingActMap,
+    ) -> Self {
+        Self {
+            once: TakeCell::new(GSState {
+                grace_period,
+                start_tasks_stream_complete,
+                cancels_tx,
+                outstanding,
+            }),
+        }
+    }
+
+    fn initiate(&self) {
+        if let Some(state) = self.once.take_once() {
+            let grace_deadline = Instant::now() + state.grace_period;
+            // We don't really need to join this
+            tokio::task::spawn(async move {
+                // Don't issue cancels until we know we won't be getting any more tasks to avoid
+                // missing the cancellation of any just-starting activities.
+                state.start_tasks_stream_complete.cancelled().await;
+                // Make sure we've waited at least the grace period. This way if waiting for
+                // starts to finish took a while, we subtract that from the grace period.
+                tokio::time::sleep_until(grace_deadline.into()).await;
+                for mapref in state.outstanding.iter() {
```

Comment:
Also, don't need the map here since you have it in the merged starts and cancels stream above.